File: Server/src/main/java/Server/core/game/node/entity/player/link/diary/MysteriousStatueZone.java
package core.game.node.entity.player.link.diary;

import core.game.node.entity.Entity;
import core.game.node.entity.player.Player;
import core.game.world.map.Location;
import core.game.world.map.zone.MapZone;
import core.game.world.map.zone.ZoneBorders;
import core.game.world.map.zone.ZoneBuilder;
import core.plugin.Initializable;
import core.plugin.Plugin;
import core.tools.Vector3d;

@Initializable
public class MysteriousStatueZone extends MapZone implements Plugin<Object> {
    private final static Vector3d origin = new Vector3d(2740.5, 3490.5, 0);
    private final static Vector3d n = new Vector3d(0, 0, 1);

    public MysteriousStatueZone() {
        super("mysterious-statue", true);
    }

    @Override
    public boolean enter(Entity e) {
        if (e instanceof Player) {
            //System.out.println("enter");
        }
        return true;
    }

    @Override
    public boolean leave(Entity e, boolean logout) {
        if (e instanceof Player) {
            //System.out.println("leave");
            e.asPlayer().removeAttribute("diary:seers:statue-walk-start");
        }
        return true;
    }

    @Override
    public void locationUpdate(Entity e, Location last) {
        if (e instanceof Player) {
            Player player = e.asPlayer();
            //System.out.println("update - " + player.getLocation().toString());
            if (!player.getWalkingQueue().isRunning() && !player.getSettings().isRunToggled()) {
                if (player.getAttribute("diary:seers:statue-walk-start") != null) {
                    Vector3d start = player.getAttribute("diary:seers:statue-walk-start");
                    Vector3d a = player.getAttribute("diary:seers:statue-walk-a");
                    Vector3d b = new Vector3d(player.getLocation()).sub(origin);

                    double angle_a_b = Vector3d.signedAngle(a, b, n) * 360. / 2 / 3.14159265355;

                    /* player.getWalkingQueue().isMoving() doesn't work right, or at least does not do what name implies it should
                    if (!player.getWalkingQueue().isMoving()) {
                        System.out.println("removing, not moving");
                        player.removeAttribute("diary:seers:statue-walk-start");
                    }*/
                    if (angle_a_b >= 0) {
                        //System.out.println("removing, not going clockwise");
                        player.removeAttribute("diary:seers:statue-walk-start");
                    }
                    if (b.epsilonEquals(start, .001)) {
                        player.getAchievementDiaryManager().finishTask(player, DiaryType.SEERS_VILLAGE, 0, 1);
                        //System.out.println("removing, finished task");
                        player.removeAttribute("diary:seers:statue-walk-start");
                    }

                    player.setAttribute("diary:seers:statue-walk-a", b);
                } else {
                    //System.out.println("started");
                    Vector3d start = new Vector3d(player.getLocation()).sub(origin);
                    player.setAttribute("diary:seers:statue-walk-start", start);
                    player.setAttribute("diary:seers:statue-walk-a", start);
                }
            } else {
                //System.out.println("removing, running or outside");
                player.removeAttribute("diary:seers:statue-walk-start");
            }
        }
    }

    @Override
    public void configure() {
        register(new ZoneBorders(2739, 3489, 2742, 3492, 0));
    }

    @Override
    public Plugin<Object> newInstance(Object arg) throws Throwable {
        ZoneBuilder.configure(this);
        return this;
    }

    @Override
    public Object fireEvent(String identifier, Object... args) {
        return null;
    }
}

File: Server/src/main/kotlin/rs09/game/node/entity/player/diary/seers/MysteriousStatueZone.kt
package rs09.game.node.entity.player.diary.seers

import core.game.node.entity.Entity
import core.game.node.entity.player.Player
import core.plugin.Initializable
import core.game.world.map.zone.MapZone
import core.tools.Vector3d
import core.game.node.entity.player.link.diary.DiaryType
import core.game.world.map.Location
import core.game.world.map.zone.ZoneBorders
import core.game.world.map.zone.ZoneBuilder
import core.plugin.Plugin
import rs09.game.content.activity.blastfurnace.BlastFurnace

@Initializable
class MysteriousStatueZone : MapZone("mysterious-statue", true), Plugin<Any?> {

    val northWest = Location.create(2739, 3492, 0)
    val northEast = Location.create(2742, 3492, 0)
    val southEast = Location.create(2742, 3489, 0)
    val southWest = Location.create(2739, 3489, 0)


    override fun newInstance(arg: Any?): MysteriousStatueZone {
        ZoneBuilder.configure(this)
        return this
    }

    override fun configure() {
        super.register(ZoneBorders(2739,3492,2742,3489))
    }

    override fun enter(e: Entity?): Boolean {
        return super.enter(e)
    }

    override fun fireEvent(identifier: String?, vararg args: Any?): Any {
        return Unit
    }

    override fun locationUpdate(e: Entity?, last: Location?) {
        if (e is Player && !e.isArtificial) {
            val player = e.asPlayer()
            if(!player.achievementDiaryManager.hasCompletedTask(DiaryType.SEERS_VILLAGE,0,1 ) && player.getAttribute("seersStatueProgress",-1) !in 0..3 ){
                player.setAttribute("seersStatueProgress",0)
                println("statue started")
            }else{
                when(player.location){
                    northWest ->{
                        if(player.getAttribute("seersStatueProgress",-1) == 0){
                            player.setAttribute("seersStatueProgress",1)
                            println("location 1 of 4 finished")
                            println(player.getAttribute("seersStatueProgress",-1))
                        }else if(player.getAttribute("seersStatueProgress",-1) in 2..3){
                            player.setAttribute("seersStatueProgress",0)
                            println("location progress reset")
                        }
                    }
                    northEast ->{
                        if(player.getAttribute("seersStatueProgress",0) == 1){
                            player.setAttribute("seersStatueProgress",2)
                            println("location 2 of 4 finished")
                        }else if(player.getAttribute("seersStatueProgress",0) != 1){
                            player.setAttribute("seersStatueProgress",0)
                            println("location progress reset")
                        }
                    }
                    southEast ->{
                        if(player.getAttribute("seersStatueProgress",0) == 2){
                            player.setAttribute("seersStatueProgress",3)
                            println("location 3 of 4 finished")
                        }else if(player.getAttribute("seersStatueProgress",0) != 2){
                            player.setAttribute("seersStatueProgress",0)
                            println("location progress reset")
                        }
                    }
                    southWest ->{
                        if(player.getAttribute("seersStatueProgress",0) == 3){
                            player.setAttribute("seersStatueProgress",4)
                            println("location 4 of 4 finished")
                        }else if(player.getAttribute("seersStatueProgress",0) != 3){
                            player.setAttribute("seersStatueProgress",0)
                            println("location progress reset")
                        }
                    }
                }
            }
            if(player.getAttribute("seersStatueProgress",0) == 4){
                player.achievementDiaryManager.finishTask(player,DiaryType.SEERS_VILLAGE,0,1)
                player.setAttribute("seersStatueComplete",true)
                player.removeAttribute("seersStatueProgress")
            }
    }
        super.locationUpdate(e, last)
    }

    override fun leave(e: Entity?, logout: Boolean): Boolean {
        if(e is Player && !e.isArtificial){
            val player = e.asPlayer()
            if(!player.getAttribute("seersStatueComplete",false) && player.getAttribute("seersStatueProgress",-0) in 0..3){
                player.removeAttribute("seersStatueProgress")
                super.leave(e, logout)
            }else return super.leave(e, logout)
        }
        return super.leave(e, logout)
    }

}